File: packages/go/internal/seguranca/ataques/ataque_separacao_chave_iv.go
package ataques

import (
	"bytes"
	"fmt"

	"criptografia/internal/seguranca"
)

// OpcoesSeparacaoChaveIV configura o AtaqueSeparacaoChaveIV.
type OpcoesSeparacaoChaveIV struct {
	Tentativas   int
	TamanhoBloco int
}

// AtaqueSeparacaoChaveIV testa se o keystream depende de key e iv apenas pela
// combinação key XOR iv: keystream(key, iv) == keystream(key^d, iv^d).
type AtaqueSeparacaoChaveIV struct {
	tentativas   int
	tamanhoBloco int
}

// NovoAtaqueSeparacaoChaveIV cria o ataque. Sem argumento, usa tentativas = 50
// e tamanhoBloco = 32.
func NovoAtaqueSeparacaoChaveIV(opcoes ...OpcoesSeparacaoChaveIV) *AtaqueSeparacaoChaveIV {
	o := OpcoesSeparacaoChaveIV{Tentativas: 50, TamanhoBloco: 32}
	if len(opcoes) > 0 {
		o = opcoes[0]
	}
	return &AtaqueSeparacaoChaveIV{tentativas: o.Tentativas, tamanhoBloco: o.TamanhoBloco}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueSeparacaoChaveIV) Nome() string {
	return "Separação chave/IV (invariância a key XOR iv)"
}

// Executar roda o ataque.
func (a *AtaqueSeparacaoChaveIV) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	tamChave := len(alvo.ChaveDeTeste())
	tamIv := alvo.TamanhoIv()

	confirmacoes := 0
	for t := 0; t < a.tentativas; t++ {
		chave := seguranca.RandomBytes(tamChave)
		iv := seguranca.RandomBytes(tamIv)
		d := seguranca.RandomBytes(tamIv)

		chave2 := make([]byte, tamChave)
		for p := 0; p < tamChave; p++ {
			chave2[p] = chave[p] ^ d[p%tamIv]
		}
		iv2 := make([]byte, tamIv)
		for p := 0; p < tamIv; p++ {
			iv2[p] = iv[p] ^ d[p]
		}

		ks1 := alvo.GerarKeystreamBruto(chave, iv, "enc", a.tamanhoBloco)
		ks2 := alvo.GerarKeystreamBruto(chave2, iv2, "enc", a.tamanhoBloco)

		if bytes.Equal(ks1, ks2) {
			confirmacoes++
		}
	}

	invariante := confirmacoes == a.tentativas

	severidade := seguranca.SevInfo
	detalhes := fmt.Sprintf(
		"Invariância não se confirmou (%d/%d); key e iv entram de forma independente.",
		confirmacoes, a.tentativas,
	)
	if invariante {
		severidade = seguranca.SevMedia
		detalhes = fmt.Sprintf(
			"Confirmado em %d/%d: keystream(key,iv) == keystream(key^d, iv^d). "+
				"O IV só desloca a chave por XOR antes da difusão - não injeta entropia independente no key schedule.",
			a.tentativas, a.tentativas,
		)
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: invariante,
		Severidade: severidade,
		Detalhes:   detalhes,
		Dados:      map[string]any{"confirmacoes": confirmacoes, "tentativas": a.tentativas},
	}, nil
}

File: src/Core/Seguranca/Ataques/AtaqueSeparacaoChaveIV.php
<?php

namespace Application\Core\Seguranca\Ataques;

use Application\Core\Seguranca\AlvoCriptografico;
use Application\Core\Seguranca\AtaqueInterface;
use Application\Core\Seguranca\ResultadoAtaque;
use Application\Core\Seguranca\SkipAtaqueException;

/**
 * Testa se o keystream depende de key e iv APENAS pela combinação key XOR iv.
 * Nessa cifra o estado inicial é `a[pos] = key[pos] ^ iv[pos]`, então, se o
 * resto do gerador não reintroduzir key/iv separadamente, vale exatamente:
 *
 *     keystream(key, iv) == keystream(key ^ d, iv ^ d)
 *
 * para qualquer máscara d (com o devido alinhamento key/iv). Isso é uma
 * propriedade estrutural relevante: significa que key e iv não entram de
 * forma independente na cifra, o que enfraquece o modelo de segurança (o IV
 * deveria contribuir com entropia própria, não só deslocar a chave por XOR).
 */
class AtaqueSeparacaoChaveIV implements AtaqueInterface
{
    public function __construct(private int $tentativas = 50, private int $tamanhoBloco = 32)
    {
    }

    public function nome(): string
    {
        return 'Separação chave/IV (invariância a key XOR iv)';
    }

    public function executar(AlvoCriptografico $alvo): ResultadoAtaque
    {
        $tamChave = strlen($alvo->chaveDeTeste());
        $tamIv = $alvo->tamanhoIv();

        $primeiro = $alvo->gerarKeystreamBruto($alvo->chaveDeTeste(), random_bytes($tamIv), 'enc', $this->tamanhoBloco);
        if ($primeiro === null) {
            throw new SkipAtaqueException('Alvo não expõe gerarKeystreamBruto.');
        }

        $confirmacoes = 0;
        for ($t = 0; $t < $this->tentativas; $t++) {
            $chave = random_bytes($tamChave);
            $iv = random_bytes($tamIv);
            $d = random_bytes($tamIv);

            $chave2 = '';
            for ($p = 0; $p < $tamChave; $p++) {
                $chave2 .= chr(ord($chave[$p]) ^ ord($d[$p % $tamIv]));
            }
            $iv2 = '';
            for ($p = 0; $p < $tamIv; $p++) {
                $iv2 .= chr(ord($iv[$p]) ^ ord($d[$p]));
            }

            $ks1 = $alvo->gerarKeystreamBruto($chave, $iv, 'enc', $this->tamanhoBloco);
            $ks2 = $alvo->gerarKeystreamBruto($chave2, $iv2, 'enc', $this->tamanhoBloco);

            if ($ks1 === $ks2) {
                $confirmacoes++;
            }
        }

        $invariante = $confirmacoes === $this->tentativas;

        return new ResultadoAtaque(
            $this->nome(),
            vulneravel: $invariante,
            severidade: $invariante ? 'media' : 'info',
            detalhes: $invariante
                ? "Confirmado em {$this->tentativas}/{$this->tentativas}: keystream(key,iv) == keystream(key^d, iv^d). " .
                  'O IV só desloca a chave por XOR antes da difusão - não injeta entropia independente no key schedule.'
                : "Invariância não se confirmou ($confirmacoes/{$this->tentativas}); key e iv entram de forma independente.",
            dados: ['confirmacoes' => $confirmacoes, 'tentativas' => $this->tentativas],
        );
    }
}
